CRM/Member/BAO/Membership.php

```diff
@@ -1594,17 +1573,35 @@ public static function buildMembershipTypeValues(&$form, $membershipTypeID = arr
       'max_related',
       'duration_unit',
       'duration_interval',
-    );
+    ];
 
-    while ($dao->fetch()) {
-      $membershipTypeValues[$dao->id] = array();
-      foreach ($membershipTypeFields as $mtField) {
-        $membershipTypeValues[$dao->id][$mtField] = $dao->$mtField;
-      }
+    $membershipTypeParams = [
+      'is_active' => $activeOnly,
+      'return' => $membershipTypeFields,
+      'options' => ['limit' => 0],
+    ];
+    if ($financialTypes) {
+      $membershipTypeParams['financial_type_id'] = ['IN' => array_keys($financialTypes)];
+    }
+    if (!empty($membershipTypeIDS)) {
+      $membershipTypeParams['id'] = ['IN' => $membershipTypeIDS];
+    }
+    if (!empty($recurOnly)) {
+      // Can be 1:optional or 2:required
+      $membershipTypeParams['auto_renew'] = ['>' => 0];
     }
 
+    $membershipTypesApi = civicrm_api3('MembershipType', 'get', $membershipTypeParams);
```

Comment:
values is a given so don't need the line below. Can just do 
```
$membershipTypeValues = civicrm_api3('MembershipType', 'get', $membershipTypeParams)['values'];
```

Reply:
@eileenmcnaughton If there were no membership types meeting the criteria (eg. none active), though not very likely it's very possible and easy to achieve via the UI.  And then `values` would not be defined causing PHP notices on the foreach.

Reply:
@mattwire - values would still be present - it would just be empty

Reply:
ok